main.go

```diff
@@ -4,191 +4,130 @@
 )
 
 const (
 	maxRetries = 5
 	sleepTime  = 10 * time.Second
 )
 
-var (
-	socketPath = "/var/run/dpdk/rte/dpdk_telemetry.v2"
-
-	promMetrics = prometheus.NewGaugeVec(
-		prometheus.GaugeOpts{
-			Name: "dpdk_interface_stat",
-			Help: "DPDK interface statistic",
-		},
-		[]string{"interface", "stat_name"},
-	)
-
-	promMetricsCallCount = prometheus.NewGaugeVec(
-		prometheus.GaugeOpts{
-			Name: "dpdk_graph_stat",
-			Help: "Dp-Service graph statistics",
-		},
-		[]string{"node_name", "graph_node"},
-	)
-)
-
-type EthdevList struct {
-	Value []int `json:"/ethdev/list"`
-}
-
-type EthdevInfo struct {
-	Value struct {
-		Name string `json:"name"`
-	} `json:"/ethdev/info"`
-}
-
-type EthdevXstats struct {
-	Value map[string]float64 `json:"/ethdev/xstats"`
-}
+func main() {
+	var conn net.Conn
+	var err error
+	var host string
+	var hostnameFlag string
+	var pollIntervalFlag int
+	var exporterPort uint64
+	var exporterAddr netip.AddrPort
 
-type DpServiceNatPort struct {
-	Value map[string]int `json:"/dp_service/nat/used_port_count"`
-}
+	log := logrus.New()
+	log.Formatter = new(logrus.JSONFormatter)
 
-type NodeData map[string]float64
+	r := prometheus.NewRegistry()
+	r.MustRegister(metrics.InterfaceStat)
+	r.MustRegister(metrics.CallCount)
 
-type GraphCallCount struct {
-	Node_0_to_255 NodeData `json:"Node_0_to_255"`
-}
+	http.Handle("/metrics", promhttp.HandlerFor(r, promhttp.HandlerOpts{}))
 
-type DpServiceGraphCallCount struct {
-	GraphCallCnt GraphCallCount `json:"/dp_service/graph/call_count"`
-}
+	conn = connectToDpdkTelemetry(log)
+	defer conn.Close()
 
-func flushSocket(conn net.Conn) {
-	respBytes := make([]byte, 1024)
+	flag.StringVar(&hostnameFlag, "hostname", "", "Hostname to use")
+	flag.IntVar(&pollIntervalFlag, "poll-interval", 20, "Polling interval in seconds")
+	flag.Uint64Var(&exporterPort, "port", 8080, "Port on which exporter will be running.")
+	flag.Parse()
 
-	_, err := conn.Read(respBytes)
-	if err != nil {
-		log.Fatalf("Failed to read response from %s: %v", socketPath, err)
+	if exporterPort < 1024 || exporterPort > 65535 {
+		log.Fatal("port must be in range 1024 - 65535")
 	}
-}
+	exporterAddr = netip.AddrPortFrom(netip.IPv4Unspecified(), uint16(exporterPort))
 
-func queryTelemetry(conn net.Conn, command string, response interface{}) {
-	_, err := conn.Write([]byte(command))
+	host, err = getHostname(hostnameFlag)
 	if err != nil {
-		log.Fatalf("Failed to send command to %s: %v", socketPath, err)
+		log.Fatal("could not get hostname")
 	}
+	log.Infof("Hostname: %s", host)
 
-	respBytes := make([]byte, 1024*6)
-	var responseBuffer bytes.Buffer
-	for {
-		n, err := conn.Read(respBytes)
-		if err != nil {
-			log.Fatalf("Failed to read response from %s: %v", socketPath, err)
-		}
-		responseBuffer.Write(respBytes[:n])
-		parts := strings.SplitN(command, ",", 2)
-		command = parts[0]
-		if bytes.Contains(respBytes, []byte(command)) {
-			break
+	go func() {
+		for {
+			if !testDpdkConnection() {
```

Comment:
I was testing the new functionality of the "reconnect" in case dpservice gets restarted.
It works as expected, if the dpservice is down and gets restarted in a time period which is longer than  `time.Duration(pollIntervalFlag) * time.Second`. If it is shorter than that (dpservice gets stopped/started quickcly)
I get such an output. Can you also detect this error and try to reconnect to dpservice ?



> sudo ./bin/prometheus-dpdk-exporter 
{"level":"info","msg":"Hostname: guvenc-vm","time":"2024-04-11T19:15:44+02:00"}
{"level":"info","msg":"Server starting on :8080...","time":"2024-04-11T19:15:44+02:00"}

> {"level":"error","msg":"Failed to send command to /var/run/dpdk/rte/dpdk_telemetry.v2: write unixpacket @-\u003e/var/run/dpdk/rte/dpdk_telemetry.v2: write: broken pipe","time":"2024-04-11T19:16:04+02:00"}
{"level":"error","msg":"Failed to send command to /var/run/dpdk/rte/dpdk_telemetry.v2: write unixpacket @-\u003e/var/run/dpdk/rte/dpdk_telemetry.v2: write: broken pipe","time":"2024-04-11T19:16:04+02:00"}
{"level":"error","msg":"Failed to send command to /var/run/dpdk/rte/dpdk_telemetry.v2: write unixpacket @-\u003e/var/run/dpdk/rte/dpdk_telemetry.v2: write: broken pipe","time":"2024-04-11T19:16:04+02:00"}
{"level":"error","msg":"Failed to send command to /var/run/dpdk/rte/dpdk_telemetry.v2: write unixpacket @-\u003e/var/run/dpdk/rte/dpdk_telemetry.v2: write: broken pipe","time":"2024-04-11T19:16:04+02:00"}

Reply:
I reworked the testing function and now it should handle also this short interruption without error.